README.md

# AttrMap

`attrmap` is an open source tool with **read-only protection** to help users get rid of dictionary in Python.

## Introduction

`attrmap` maps the `dict` to an object of `AttrMap`, so users can avoid using the annoying `[""]` syntax.

Users are welcome to report issues or request new features.

The API references can be found [here](https://attrmap.readthedocs.io/en/latest/index.html).

## What's New

- 2022-12-20: Fix the annoying warning when creating an AttrMap instance.
- 2022-08-27: Add utilities for AttrMap, see [docs](https://attrmap.readthedocs.io/en/latest/index.html).
- 2022-08-27: Add deprecated warnings.
- 2022-08-13: Fix the `delattr` bug, users can delete an item via `del obj.attr` when the `AttrMap` object is modifiable.
- 2022-08-13: Fix the improper use of type hint, now `attrmap` is available for python>=3.6.
- 2022-08-13: Update the document.

## Installation

You can install `attrmap` via pip:

```bash
pip install -U attrmap
```

## Getting Started

Assuming you have an instance of `dict`, then you can build an object of `AttrMap` as follows:

```python
from attrmap import AttrMap
import attrmap.utils as au

CONFIGS = {
    "attr1": 1, 
    "attr2": ["hello", " ", "world"], 
    "attr3": {
        "subattr1": "subattr1", 
        "subattr2": {
            "subsubattr1": "subsubattr1",
        }
    }
}

configs = AttrMap(CONFIGS)

# Set the object unmodifiable.
configs = au.convert_state(configs, read_only=True)
# Equivalent to:
# au.convert_state(configs, read_only=True)

print(configs) # The outputs is more human readable.
# Object Contains Following Attributes
#  attr1: 1
#  attr2: ['hello', ' ', 'world']
#  attr3:
#          subattr1: subattr1
#          subattr2:
#                  subsubattr1: subsubattr1
```

You can also assign the attributes and their values in a more simple way:

```python
# Assume you have import AttrMap
# Create an empty AttrMap object.
configs = AttrMap()
configs.attr1 = 1
configs.attr2 = ["hello", " ", "world"]
configs.attr3.subattr1 = "subattr1"
configs.attr3.subattr2.subsubattr1 = "subsubattr1"

print(configs)
# Object Contains Following Attributes
#  attr1: 1
#  attr2: ['hello', ' ', 'world']
#  attr3:
#          subattr1: subattr1
#          subattr2:
#                  subsubattr2: subsubattr1
```

You can convert the `AttrMap` object to Python `dict` anytime:

```python
configs_dict = au.todict(configs)

print(configs_dict)
# {'attr1': 1, 'attr2': ['hello', ' ', 'world'], 'attr3': {'subattr1': 'subattr1', 'subattr2': {'subsubattr1': 'subsubattr1'}}}

print(type(configs_dict))
# <class 'dict'>

# AttrMap object doesn't store the original dict. 
print(id(configs_dict))
# 139881853477376
print(id(CONFIGS))
# 139881853478464

# But it do not change the items:
print(configs_dict == CONFIGS)
# True
```

`attrmap` has the `read-only` protection to protect you from unintentional modification:

```python
# Assume you have the above AttrMap object named configs
# Now set the AttrMap object as read-only.
configs = au.convert_state(configs, read_only=True)

# Try to create a new attribute.
configs.attr4 = "unintentional modification"
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
#   File "/xxxx/attrmap.py", line 558, in __setattr__
#     self._check_modifiable()
#   File "/xxxx/attrmap.py", line 643, in _check_modifiable
#     raise AttributeError(
# AttributeError: Modifying the attributes of a read-only AttrMap instance is not allowed.

# Try to modify the value of existing attribute:
configs.attr1 = "unintentional modification"
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
#   File "/xxxx/attrmap.py", line 558, in __setattr__
#     self._check_modifiable()
#   File "/xxxx/attrmap.py", line 643, in _check_modifiable
#     raise AttributeError(
# AttributeError: Modifying the attributes of a read-only AttrMap instance is not allowed.
```

The `read-only` protection is quite useful when you use the `AttrMap` object to store the configuration of any system, any unintentional modification is not allowed. If you want to update the attribute's value, just set the `read_only` of the `AttrMap` instance as `False` via `au.convert_state(obj, read_only=False)`.

If the `AttrMap` object should be unmodifiable, setting the object as **read-only** is recommended.

You can also build an `AttrMap` instance from a `.json` or `.yaml` file. For example, assume you have a `file.json` file:

```json
{
    "language": ["python", "cpp", "java"], 
    "value": [["one", "two"]], 
    "structure":{
        "tree":["left tree", "right tree"]
    }
}
```

Then you can build `AttrMap` object via:

```python
from attrmap import AttrMap

configs = AttrMap(path2file="/PATH_TO_YOUR_DIRECTORY/file.json") # .yaml file works too.
print(configs)
# Object Contains Following Attributes
#  language: ['python', 'cpp', 'java']
#  value: [['one', 'two']]
#  structure:
#          tree: ['left tree', 'right tree']
```

**NOTE:** The `read-only` protection should be explicitly enabled, `AttrMap` assumes you are still setting the attributes and their values.

## Known Issues

- Attributes start with `_` might be conflict with `AttrMap` perserved methods/properties. Attributes can still be accessed via dict-style (`[""]`), but the attribute-style
will return the "protected" (not the term of python, only means it's not "public") method such as `_build_from_dict`.
- Some public available properties and methods are reserved, they can only accessed via the dict-style.

## Running Tests

Open bash shell and execute:

```bash
bash -i ./auto_test.sh
```

## APIs

See [The API Reference](https://attrmap.readthedocs.io/en/latest/modules.html)

## TODO

## LICENSE

`AttrMap` is an open source tool under Apache 2.0 license.


CHANGELOG.md

# CHANGELOG

This file maintains the change log of `AttrMap`.

- 2022-12-20: Fix the annoying warning when creating an AttrMap instance.
- 2022-08-27: Add utilities for AttrMap, see [docs](https://attrmap.readthedocs.io/en/latest/index.html).
- 2022-08-27: Add deprecated warnings.
- 2022-08-13: Fix the `delattr` bug, you can delete an item via `del obj.attr` when `AttrMap` object is modifiable.
- 2022-08-13: Fix the improper use of type hint, now `attrmap` is available for python>=3.6.
- 2022-08-13: Update the document.


test_env.sh
source ~/.bashrc

min_ver=6
max_ver=10

# executing unit tests
for ((ver=$min_ver; ver <= $max_ver; ver++))
do
    echo "executing unit tests under python version $ver"
    conda activate py3.$ver
    python3 -m pip install --user -e .
    pytest tests
    conda deactivate 
    conda remove -n py3.$ver --all -y
done

auto_test.sh

min_ver=6
max_ver=10

# create conda environment.
for ((ver=$min_ver; ver <= $max_ver; ver++))
do 
    echo "creating conda environment with python version 3.$ver"
    conda create -n py3.$ver python=3.$ver -y
    conda activate py3.$ver
    conda install pytest -y
    python3 -m pip install -e .
    # pip install -U attrmap
    pytest tests
    conda deactivate
done

for ((ver=$min_ver; ver <= $max_ver; ver++))
do
    echo "removing python environment of version 3.$ver"
    conda remove -n py3.$ver --all -y
done


tests/test_utils.py

import yaml
import json
from attrmap import AttrMap
from copy import deepcopy as dcp
import pytest

from attrmap.utils import (
    contains,
    convert_state,
    get_items,
    get_keys,
    get_vals,
    is_modifiable,
    is_read_only,
    merge_from,
    todict)


CONFIGS = {
    "attr1": 1,
    "attr2": ["hello", " ", "world"],
    "attr3": {
        "subattr1": "subattr1",
        "subattr2": {
            "subsubattr1": "subsubattr1",
        }
    }
}


def test_attrmap():
    configs = AttrMap(CONFIGS)
    cfg = AttrMap(attr1=1)
    cfg.attr2 = dcp(CONFIGS["attr2"])
    cfg.attr3.subattr2.subsubattr1 = "subsubattr1"
    cfg.attr3.subattr1 = "subattr1"

    assert cfg == configs


def test_build_from_file():
    cfg_json = AttrMap(
        path2file=__file__.replace("test_utils.py", "test.json")
    )
    cfg_yaml = AttrMap(
        path2file=__file__.replace("test_utils.py", "test.yaml")
    )
    assert cfg_json == cfg_yaml


def test_merge_from():
    path2file = __file__.replace("test_utils.py", "test.yaml")
    cfg_yaml = AttrMap(path2file=path2file)
    assert "language" in get_keys(cfg_yaml), get_keys(cfg_yaml)
    configs = AttrMap(dcp(CONFIGS))
    merge_from(configs, cfg_yaml)
    for key in list(CONFIGS.keys()):
        assert key in get_keys(configs), get_keys(configs)
    for key in get_keys(cfg_yaml):
        assert key in get_keys(configs), get_keys(configs)

    with open(path2file, 'r') as fp:
        mapping_file = yaml.safe_load(fp)
    configs = AttrMap()
    merge_from(configs, path2file=path2file)
    todict(configs) == mapping_file


def test_read_only():
    configs = AttrMap(CONFIGS)
    convert_state(configs, True)
    assert configs.__dict__["READ_ONLY"] is is_read_only(configs)
    assert configs.__dict__["READ_ONLY"] is not is_modifiable(configs)
    try:
        configs.attr1 = "hello world"
        assert False, configs.__dict__["READ_ONLY"]
    except AttributeError:
        pass


def test_to_dict():
    configs = AttrMap(CONFIGS)
    configs = todict(configs)
    assert configs == CONFIGS


def test_convert_state():
    configs = AttrMap(CONFIGS)
    _configs = convert_state(configs, read_only=True)
    assert configs is _configs
    assert is_read_only(configs) is True
    convert_state(configs, read_only=False)
    assert is_read_only(configs) is False


def test_contains():
    configs = AttrMap(CONFIGS)
    for key in CONFIGS.keys():
        assert contains(configs, key), key
        assert key in configs


def test_getitem():
    configs = AttrMap(CONFIGS)
    for key in CONFIGS.keys():
        val = configs[key]
        if isinstance(val, AttrMap):
            val = todict(val)
        assert val == CONFIGS[key], val


def test_keys():
    configs = AttrMap(CONFIGS)
    assert sorted(list(CONFIGS.keys())) \
           == sorted(get_keys(configs)),\
           configs.keys()


def test_vals():
    configs = AttrMap(CONFIGS)
    attr_vals = get_vals(configs)
    dict_vals = CONFIGS.values()
    for attr_val, dict_val in zip(attr_vals, dict_vals):
        if isinstance(attr_val, AttrMap):
            attr_val = todict(attr_val)
        assert attr_val == dict_val, attr_val


def test_items():
    configs = AttrMap(CONFIGS)
    attr_items = get_items(configs)
    dict_items = CONFIGS.items()
    # dict_items = list(dict_items)
    for attr_it, dict_it in zip(attr_items, dict_items):
        assert attr_it[0] == dict_it[0], attr_it[0]
        val = attr_it[1]
        if isinstance(val, AttrMap):
            val = todict(val)
        assert val == dict_it[1], attr_it[1]


def test_delete_item():
    configs = AttrMap(CONFIGS)

    convert_state(configs, read_only=True)
    with pytest.raises(AttributeError):
        del configs.attr1
    assert hasattr(configs, "attr1")

    with pytest.raises(AttributeError):
        del configs.attr3.subattr2.subsubattr1
    assert hasattr(configs.attr3.subattr2, "subsubattr1")

    with pytest.raises(AttributeError):
        del configs.attr3.subattr1
    assert hasattr(configs.attr3, "subattr1")

    convert_state(configs, read_only=False)
    del configs.attr1
    convert_state(configs, read_only=True)
    assert not hasattr(configs, "attr1")
    convert_state(configs, read_only=False)
    del configs.attr3.subattr2.subsubattr1
    convert_state(configs, read_only=True)
    assert not hasattr(configs.attr3.subattr2, "subsubattr1")
    convert_state(configs, read_only=False)
    del configs.attr3["subattr1"]
    convert_state(configs, read_only=True)
    assert not hasattr(configs.attr3, "subattr1")


attrmap/utils.py

"""
Utilities for AttrMap.

Author:
    Yiqun Chen
"""

from copy import deepcopy
from typing import Dict, Iterable, List, Union
import os
import yaml
import json

from attrmap import AttrMap


_READ_ONLY = 'READ_ONLY'
_ATTRMAP_PREFIX = 'ATTRMAP_PREFIX'
_ATTRMAP_LEVEL = 'ATTRMAP_LEVEL'
_RESERVED = 'RESERVED'


def is_read_only(am: AttrMap) -> bool:
    """
    Return the state of `AttrMap` object.
    `True` is read-only and any modification to the object
    is not allowed, e.g., add or delete an attribute,
    change the value of any attribute.

    >>> configs = AttrMap(CONFIGS)
    >>> configs = convert_state(configs, read_only=True)
    >>> print(configs)
    Object Contains Following Attributes
        attr1: 1
        attr2: ['hello', ' ', 'world']
        attr3:
            subattr1: subattr1
            subattr2:
                    subsubattr1: subsubattr1
    # Try to create a new attribute.
    >>> configs.attr4 = "unintentional modification"
    Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "/xxxx/attrmap.py", line xxx, in __setattr__
            self._check_modifiable()
        File "/xxxx/attrmap.py", line xxx, in _check_modifiable
            raise AttributeError(
        AttributeError: A read only AttrMap instance is not allowed to modify\
            its attribute.
    # Try to modify the value of existing attribute:
    >>> configs.attr1 = "unintentional modification"
    Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "/xxxx/attrmap.py", line xxx, in __setattr__
            self._check_modifiable()
        File "/xxxx/attrmap.py", line xxx, in _check_modifiable
            raise AttributeError(
        AttributeError: A read only AttrMap instance is not allowed to modify\
            its attribute.
    >>> print(configs)
    Object Contains Following Attributes
        attr1: 1
        attr2: ['hello', ' ', 'world']
        attr3:
            subattr1: subattr1
            subattr2:
                    subsubattr1: subsubattr1
    """
    return am.__dict__[_READ_ONLY]


def is_modifiable(am: AttrMap) -> bool:
    """
    Return the state of `AttrMap` object.
    If the object is read only, then return False, and vice versa.
    """
    return not is_read_only(am)


def todict(am: AttrMap) -> Dict:
    """
    Convert to a python builtin dict, which does not
    share memory with AttrMap instance.
    The state (is read-only or not) will not influence this method.
    For example:

    >>> configs = AttrMap(CONFIGS)
    >>> configs_dict = todict(configs)
    >>> print(configs_dict)
    {'attr1': 1, 'attr2': ['hello', ' ', 'world'],
    'attr3': {'subattr1': 'subattr1',
    'subattr2': {'subsubattr1': 'subsubattr1'}}}
    >>> print(type(configs_dict))
    <class 'dict'>
    # AttrMap object doesn't store the original dict.
    >>> print(id(configs_dict))
    139881853477376
    >>> print(id(CONFIGS))
    139881853478464
    # But it do not change the items:
    >>> print(configs_dict == CONFIGS)
    True
    """
    result = {}
    _prefix = _get_prefix(am)
    for key, value in am.__dict__.items():
        value = todict(value) \
            if isinstance(value, AttrMap) else value
        if key.startswith(_prefix):
            new_key = key.replace(_prefix, "")
            result[new_key] = value
    return result


convert_to_dict = todict
to_dict = todict


def convert_state(am: AttrMap, read_only: bool) -> AttrMap:
    """
    Convert the state as read only or not.

    This method will return the self object,
    so the both two code snipts are equivalent:

    >>> configs = AttrMap(CONFIGS)
    >>> convert_state(configs, True)
    >>> configs_new = convert_state(configs, True)
    >>> id(configs) == id(configs_new)
    True
    """
    if read_only not in [True, False]:
        raise ValueError(
            f"Expect attribute read_only takes value "
            f"from [True, False, None], but got {read_only}")
    am.__dict__["READ_ONLY"] = read_only
    for _, value in am.__dict__.items():
        if isinstance(value, AttrMap):
            convert_state(value, read_only)
    return am


def convert_to_read_only(am: AttrMap) -> AttrMap:
    """
    Convert the statue to read only.
    """
    return convert_state(am, True)


def convert_to_modifiable(am: AttrMap) -> AttrMap:
    """Convert the state to modifiable."""
    return convert_state(am, False)


def merge_from(
        am: AttrMap,
        mapping: AttrMap = None,
        path2file: os.PathLike = None) -> AttrMap:
    """
    Merge from a file (json or yaml) or a mappable object (dict or AttrMap).

    .. NOTE:: The existing value will be overwrited.

    Args:
        am:
            AttrMap object, key-value pairs of other object
            will be merged in it.
        mapping:
            An instance of dict or AttrMap.
        path2file:
            The path to the json or yaml file.

    Returns:
        The merged object, shares id with input am.

    >>> configs = AttrMap(CONFIGS)
    >>> print(configs)
    Object Contains Following Attributes
    attr1: 1
    attr2: ['hello', ' ', 'world']
    attr3:
            subattr1: subattr1
            subattr2:
                    subsubattr1: subsubattr1
    >>> id(configs)
    140416566573808
    >>> configs_new = merge_from(configs, {'attr4': 'attr4'})
    >>> print(configs)
    Object Contains Following Attributes
    attr1: 1
    attr2: ['hello', ' ', 'world']
    attr3:
            subattr1: subattr1
            subattr2:
                    subsubattr1: subsubattr1
    attr4: attr4
    >>> id(configs_new) == id(configs)
    True
    >>> configs is configs_new
    True
    """
    if is_read_only(am):
        raise AttributeError("Cannot merge to a read only AttrMap object.")
    if mapping is not None:
        new = merge_from_mapping(AttrMap(), mapping)
        # Make sure the original object is modified instead creating
        # a new object.
        am.__dict__.update(new.__dict__)
    if path2file is not None:
        new = merge_from_file(AttrMap(), path2file)
        # Make sure the original object is modified instead creating
        # a new object.
        am.__dict__.update(new.__dict__)
    return am


def get_keys(am: AttrMap) -> List:
    """
    Get the (top level) keys of `AttrMap` object.

    Returns:
        keys: A list of keys.

    >>> configs = AttrMap(CONFIGS)
    >>> get_keys(configs)
    ['attr1', 'attr2', 'attr3']
    >>> type(configs.keys())
    <class 'list'>

    That is, only the top-level attribute names will be returned.
    Similarily:

    >>> get_keys(configs['attr3']) # equivalent to get_keys(configs.attr3)
    ['subattr1', 'subattr2']
    """
    _prefix = _get_prefix(am)
    keys = filter(
        lambda x: x.startswith(_prefix), am.__dict__.keys()
    )
    keys = map(lambda x: x.replace(_prefix, ""), keys)
    return list(keys)


def get_vals(am: AttrMap) -> List:
    """
    Get the (top level) values of `AttrMap` object.
    The alias of `values` is `vals`.

    Returns:
        values: A list of values.

    >>> values = get_vals(configs)
    >>> print(type(values))
    <class 'list'>
    >>> print(values)
    [1, ['hello', ' ', 'world'],     subattr1: subattr1
        subattr2:
                subsubattr1: subsubattr1, ]
    >>> type(values[0])
    <class 'int'>
    >>> type(values[1])
    <class 'list'>
    >>> type(values[2])
    <class 'attrmap.attrmap.AttrMap'>
    >>> type(values[3])
    <class 'attrmap.attrmap.AttrMap'>

    .. NOTE:: If the returned value is modified by user,
        the original AttrMap object may be modified too.
    """
    values = []
    keys = get_keys(am)
    for key in keys:
        values.append(am[key])
    return values


get_values = get_vals


def get_items(am: AttrMap) -> Iterable:
    """
    Get the (top level) `key-value` pair of `AttrMap` object like `dict`.

    Returns:
        items: zip object.

    >>> configs = AttrMap(CONFIGS)
    >>> get_items(configs)
    <zip object at 0x7fdaf3843a80>
    >>> for key, val in get_items(configs):
    ...     print(key, val)
    ... # End of for loop.
    attr1 1
    attr2 ['hello', ' ', 'world']
    attr3   subattr1: subattr1
            subattr2:
                    subsubattr1: subsubattr1
    """
    return zip(get_keys(am), get_vals(am))


def contains(am: AttrMap, key: str) -> bool:
    """
    Check if `AttrMap` contains specific attribute.

    >>> configs = AttrMap(CONFIGS)
    >>> for obj in configs:
    ...     print(obj)
    ... # End of for loop.
    ('attr1', 1)
    ('attr2', ['hello', ' ', 'world'])
    ('attr3', {'subattr1': 'subattr1', \
        'subattr2': {'subsubattr1': 'subsubattr1'}})
    """
    return key in am


def merge_from_mapping(
    mapping: Union[AttrMap, Dict],
    another: Union[AttrMap, Dict],
) -> AttrMap:
    """
    Merge another dict or AttrMap object to a given dict or AttrMap object.

    Args:
        mapping:
            The given object.
        another:
            Another object.

    Returns:
        obj: An AttrMap object.

    .. NOTE:: The value of the given object `mapping` will be overwritten.
    """
    if isinstance(mapping, AttrMap):
        mapping = todict(mapping)
    if isinstance(another, AttrMap):
        another = todict(another)
    for key, val in another.items():
        if key not in mapping.keys():
            mapping[key] = val
        elif isinstance(val, dict) \
                and isinstance(mapping[key], dict):
            cfg = todict(merge_from_mapping(mapping[key], val))
            mapping[key] = cfg
        else:
            mapping[key] = val
    return AttrMap(mapping)


def merge_from_file(
    mapping: Union[dict, AttrMap], path2file: Union[str, os.PathLike]
) -> AttrMap:
    """
    Similar to `merge_mapping`, the difference lie in another source, here
    another source is a json file or a yaml file.

    Args:
        mapping:
            The given source.
        path2file:
            The path to another source,
            currently supports both json and yaml file.

    Returns:
        obj: An AttrMap object.
    """
    if not os.path.exists(path2file):
        raise FileNotFoundError(f"File {path2file} not found.")
    with open(path2file, 'r') as fp:
        if path2file.endswith(".yaml") or path2file.endswith(".yml"):
            mapping_file = yaml.safe_load(fp)
        if path2file.endswith(".json"):
            mapping_file = json.load(fp)
    mapping = merge_from_mapping(mapping, mapping_file)
    return mapping


def _get_prefix(am: AttrMap) -> str:
    return am.__dict__[_ATTRMAP_PREFIX]


def _wrap_name(am: AttrMap, name: str) -> str:
    _prefix = _get_prefix(am)
    return f"{_prefix}{str(name)}"


def _get_level(am: AttrMap) -> int:
    return am.__dict__[_ATTRMAP_LEVEL]
